OpenRPC Support

meowsbits opened the pull request:
Rel https://github.com/filecoin-project/devgrants/pull/165

This is an implementation adding OpenRPC Document support for the Filecoin API(s). 

Note that the `docsgen` program (easily accessed via `make docsgen`) also uses reflection to generate Markdown documentation, and that program has been refactored to share some logic with the OpenRPC documentation to avoid code duplication (eg. examples and comments).

Would love to get some feedback on the general architecture (ie Is plugging directly in to the `rpc` package a best approach?), as well as a few other things, like Is re-using the given examples from the docsgen acceptable?, Is it desirable to add more detailed schemas for common data types (which has been exemplified by a tentative `cid.Cid`)? 

We'd be happy and eager to schedule a meeting to talk through some of these ideas and questions if interest and time allows. 

---

Here's a gist of what the rendered `rpc.discover` response can look like at the OpenRPC Playground: https://playground.open-rpc.org/?url=https://gist.githubusercontent.com/meowsbits/2f22b61ce46f6897d343f850ac86b540/raw/openrpc_doc.json

![image](https://user-images.githubusercontent.com/45600330/98011479-827a2700-1dbd-11eb-8488-8a7a1067b03b.png)

whyrusleeping commented:
Oooo, this is interesting. At a first glance the integration seems pretty clean, will review further tomorrow

meowsbits commented:
Hey @whyrusleeping, just bumping on this. Feedback would be very welcome.

magik6k reviewed (changes requested):
I like this idea, but there are a couple of issues

I'd strongly prefer not running any generation at runtime, to not include new / unaudited (by us) dependencies in the lotus/lotus-miner binaries

Instead of hooking into the RPC handler directly, which I don't think is needed, I'd:
* Generate the doc json blob with docgen (ok if it uses external deps, but please limit changes to deps we already include); store that blob in `build/openrpc/[full/miner].json`
* Create go-rice loaders like the one for proof parameters json file (https://github.com/filecoin-project/lotus/blob/master/build/parameters.go); as long as that call is in the `build/` package, go-rice should figure out embedding automagically
* Add a `Discover(ctx) (json.RawMessage, error)` method to CommonAPI interfaces/structs in `api/`, `api/struct/`, `node/impl/common`
  * Given that OpenRPC spec requires the method to be called `rpc.discover`, we can handle this in go-jsonrpc by adding a RPCServer.Alias method, and aliasing that method
* In the implementation just return that pre-generated blob

magik6k commented on go.sum:
```diff
@@ -212,19 +254,28 @@ github.com/dustin/go-humanize v1.0.0/go.mod h1:HtrtbFcZ19U5GC7JDqmcUSB87Iq5E25Kn
+github.com/etclabscore/go-jsonschema-walk v0.0.6/go.mod h1:VdfDY72AFAiUhy0ZXEaWSpveGjMT5JcDIm903NGqFwQ=
+github.com/etclabscore/go-openrpc-reflect v0.0.34 h1:++IsfuK+3J4s+DTWqiyj7HscojPxy3KFV4ZU8vhiQBQ=
+github.com/etclabscore/go-openrpc-reflect v0.0.34/go.mod h1:XgVhDkb6BeOtc4qf04k4K6lNI5gpcEYrAPgB8GGnIJc=
+github.com/ethereum/go-ethereum v1.9.12 h1:EPtimwsp/KGDSiXcNunzsI4kefdsMHZGJntKx3fvbaI=
+github.com/ethereum/go-ethereum v1.9.12/go.mod h1:PvsVkQmhZFx92Y+h2ylythYlheEDt/uBgFbl61Js/jo=
```
(go-ethereum being LGPL is the reason why we've implemented go-jsonrpc in the first place)

meowsbits replied:
https://github.com/filecoin-project/lotus/pull/4711/commits/413d02fab5110bdcfa6bec4f0e584d82accdd053

meowsbits commented:
Hey @magik6k! Thanks for your thoughtful and helpful review.

- I've removed the `github.com/ethereum/go-ethereum` dependency (here, and in `github.com/etclabscore/go-openrpc-reflect` as well);
  it was used only for an `Example` tests in that library.

- I've proposed an addition of method `RPCServer.AliasMethod` to `github.com/filecoin-project/go-jsonrpc` at https://github.com/filecoin-project/go-jsonrpc/pull/37,
  enabling the aliasing of `Filecoin.Discover` to `rpc.discover`, per the spec.
  Please note the commit message there mentioning potential improvement w/r/t json rpc
  "system extension" handling. _Please note that until that is merged, you'll need to `replace` that dep._ 

- The current documentation-generating commands' source trees are structured as follows. 
  
  ```
    > tree api/docgen api/docgen-openrpc
    api/docgen
    ├── cmd
    │   └── docgen.go
    └── docgen.go
    api/docgen-openrpc
    ├── cmd
    │   └── docgen_openrpc.go
    └── openrpc.go
  ```

  With the `Makefile` commands as:
  
  ```
    docsgen: docsgen-documentation-md docsgen-openrpc-json
    
    docsgen-documentation-md:
    	go run ./api/docgen/cmd "api/api_full.go" "FullNode" > documentation/en/api-methods.md
    	go run ./api/docgen/cmd "api/api_storage.go" "StorageMiner" > documentation/en/api-methods-miner.md
    	go run ./api/docgen/cmd "api/api_worker.go" "WorkerAPI" > documentation/en/api-methods-worker.md
    
    docsgen-openrpc-json:
    	go run ./api/docgen-openrpc/cmd "api/api_full.go" "FullNode" > build/openrpc/full.json
    	go run ./api/docgen-openrpc/cmd "api/api_storage.go" "StorageMiner" > build/openrpc/miner.json
    	go run ./api/docgen-openrpc/cmd "api/api_worker.go" "WorkerAPI" > build/openrpc/worker.json
  ```
  
- I added the `Discover` method not only to `*CommonAPI`, but also to `*StorageMiner` and `*worker` APIs. BUT, I'm not sure
  this is actually going to work as desired. I've done a little digging around the docs, but
  haven't found the default HTTP port (if any) for the `lotus-miner` with which I can do a basic
  sanity check. __PTAL?__

meowsbits commented on cmd/lotus/rpc.go:
```diff
@@ -32,9 +35,24 @@ import (
+
+	doc.RegisterReceiverName("Filecoin", fullAPI)
+
+	rpcServer.Register("rpc", &RPCDiscovery{doc})
```
Removed this approach with https://github.com/filecoin-project/lotus/pull/4711/commits/a63eeef4e6903ace8bc67282e0a509a49b29161a

magik6k commented on build/openrpc.go:
```diff
@@ -0,0 +1,30 @@
+type OpenRPCDocument map[string]interface{}
+
+func OpenRPCDiscoverJSON_Full() OpenRPCDocument {
+	data := rice.MustFindBox("openrpc").MustBytes("full.json")
+	m := OpenRPCDocument{}
+	_ = json.Unmarshal(data, &m)
+	return m
+}
+
+func OpenRPCDiscoverJSON_Miner() OpenRPCDocument {
+	data := rice.MustFindBox("openrpc").MustBytes("miner.json")
+	m := OpenRPCDocument{}
+	_ = json.Unmarshal(data, &m)
+	return m
+}
+
+func OpenRPCDiscoverJSON_Worker() OpenRPCDocument {
+	data := rice.MustFindBox("openrpc").MustBytes("worker.json")
```
Given that those are large-ish (1MiB), I'd gzip those files before embedding into the binary (not that 1MiB is a lot with our 100M+ binaries, but still worth trying to save some space)

meowsbits replied:
This was addressed with https://github.com/filecoin-project/lotus/pull/4711/commits/145e8bb4125c72fc6cbdbf604e8e404fa38c44aa and the couple subsequent commits.

magik6k commented on api/docgen-openrpc/cmd/docgen_openrpc.go:
```diff
@@ -0,0 +1,49 @@
+		log.Fatalln(err)
+	}
+
+	jsonOut, err := json.MarshalIndent(out, "", "    ")
+	if err != nil {
+		log.Fatalln(err)
+	}
```
I'd make this non-pretty and gzip to save space in binaries. Users can just do `zcat file.json.gz | jq` to get a readable thing

meowsbits replied:
See https://github.com/filecoin-project/lotus/pull/4711#discussion_r560397063

magik6k reviewed (commented):
Still some dep changes we need to look through, but overall no major comments, looks pretty close to landing

(merged the go-jsonrpc PR)

meowsbits commented:
It seems [the failing test](https://app.circleci.com/pipelines/github/filecoin-project/lotus/12446/workflows/8b59a1e4-45d0-4cf8-99ba-9cf31aa5e46a/jobs/107225) is not related to this PR... :crossed_fingers:

magik6k commented:
Merged in https://github.com/filecoin-project/lotus/pull/5843
